### PR TITLE
Implement MetalMachine Controller

### PR DESCRIPTION
**Pull request by defo89**
# Proposed Changes

-
-
-

Fixes #

### CODE REVIEW

File: internal/controller/metalmachine_controller.go [hunk 1]
```diff
@@ -35,11 +42,101 @@ type MetalMachineReconciler struct {
...
+	// Always close the scope when exiting this function, so we can persist any MetalMachine changes.
+	defer func() {
+		if err := machineScope.Close(); err != nil && err == nil {
+			logger.Error(err, "failed to close MetalMachine scope")
+		}
+	}()
 
-	// TODO(user): your logic here
+	// Return early if the object or Cluster is paused.
+	if annotations.IsPaused(cluster, metalMachine) {
```
Comment (hardikdr) [inline internal/controller/metalmachine_controller.go:128]: Let's discuss offline if it's possible to have a self-sufficient `MetalCluster`, `MetalMachine` API that can be deployed/tested independently of the `Machine`, `Cluster`, mainly to achieve the separation of concerns here.

**Review by hardikdr — COMMENTED** [review]
Thanks a lot @defo89 for the PR !! Few comments inline, looks good otherwise.

File: internal/controller/metalmachine_controller.go [hunk 2]
```diff
@@ -35,11 +42,101 @@ type MetalMachineReconciler struct {
...
+	logger = logger.WithValues("cluster", klog.KObj(cluster))
+
+	metalClusterName := client.ObjectKey{
+		Namespace: metalMachine.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}
+
+	metalCluster := &infrav1.MetalCluster{}
+	if err := r.Client.Get(ctx, metalClusterName, metalCluster); err != nil {
+		logger.Info("MetalCluster is not available yet")
```
[the remark this answers is not in the document]
Comment (defo89) [reply internal/controller/metalmachine_controller.go]: thanks, I added explicit handling of NotFound and Status.Ready==false

File: internal/controller/metalmachine_controller.go [hunk 3]
```diff
@@ -35,11 +42,101 @@ type MetalMachineReconciler struct {
...
+		MetalMachine: metalMachine,
+	})
+
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create machine scope: %+v", err)
+	}
+
+	// Always close the scope when exiting this function, so we can persist any MetalMachine changes.
+	defer func() {
+		if err := machineScope.Close(); err != nil && err == nil {
```
[the remark this answers is not in the document]
Comment (defo89) [reply internal/controller/metalmachine_controller.go]: lets see, I have added the suggestion and added todo

**Review by hardikdr — APPROVED** [review]
Looks good, thanks.
Needs rebase :)

**Comment by defo89** [comment]
> Looks good, thanks. Needs rebase :)

ack, rebased the PR

**Pull request merged**